README.md
# pizmidi
Simple and free MIDI plug-ins based on VST2 SDK and pizmidi framework/examples.
Without dedicated GUI, but parameters can get edited from the VST host.

Currently includes:
  * midiUnifyChannel (a simple template for new modules)
  * midiProgramChange for Arturia MIDI keyboards
  * midiFromJoystick (translates XInput to MIDI)
  * midiUartBridge for Arduino

## midiUnifyChannel 
A simple template for new VST plug-ins. 
Maps all incoming MIDI messages to the selected channel.

## midiProgramChange for Arturia MIDI keyboards
This is a program change VST plug-in for Arturia MIDI keyboards. 
While their MIDI keyboards allow user configurations, it seems that it is not possible to generate standard MIDI program changes.

You might be able to remap MIDI events in your VST Host (DAW), but as the "Preset" knob only sends relative changes, it is not so trivial.
This VST plug-in maps the default Arturia MIDI CC commands of the "Preset" knob to MIDI program changes while keeping track of the absolute program number.

How to use/operate:
  * rotating the "Preset" knob sends program changes instead
  * pressing down the "Preset" knob changes to the first program (reset)
  * all other MIDI events get passed thru

No configuration needed, no GUI, just wire in your VST host this VST plug-in after the Arturia MIDI keyboard.

But what if 128 programs are not enough for you? You need bank select commands. Fortunately, the Arturia MIDI keyboard has a "Category" knob.
So this VST plug-in maps also the default Arturia MIDI CC commands of the "Category" knob to MIDI bank select commands (7-bit LSB) while keeping track of the absolute bank number.

(The expected MIDI CC ID for CategoryEnc/CategoryBtn/PresetEnc/PresetBtn are 112/113/114/115. This is the default configuration at least on my Arturia Laboratory MIDI keyboard.)

## midiFromJoystick (translates XInput to MIDI)
Use your Joystick/Gamepad/Game controller to generate MIDI events (Note On/Off, Pitch Bench, Sustain, Portamento, Expression Controller, Program Change, Bank Select),
directly in your DAW / VST host.

Proof-of-concept for Xbox360 controllers (XInput) with a hardwired mapping: 
  * play notes with buttons A/B/X/Y plus the DPAD (8 notes from C4 to C5)
  * pitch bend with left thumb stick Y-axis
  * program changes with left shoulder / right shoulder / start button  (to decrement/increment/reset)
  
If your Joystick/Gamepad/Game controller is not XInput compatible, use the free and configurable XOutput tool.

## midiUartBridge for Arduino
This is a MIDI <-> USB <-> UART bridge implemented as VST2 plug-in.
A more robust alternative to tools like the commonly used "Hairless MIDI to Serial Bridge".

Using a virtual UART over USB, as typically used to program and communicate with Arduino Uno and similar devices,
it allows to send and receive MIDI events over USB, without any MIDI circuit/interface on the Arduino.

A simple example for an Arduino Uno, working out of the box, is provided [here](doc/ArduMidiTest.ino).
The example works also on Arduino Leonardo compatible boards (e.g. Pro Micro), but since they implement USB-MIDI
directly, the midiUartBridge or other hairy tools are not needed...

## Download / install / use
Download the Windows 10 VST2 plug-ins as either 32-bit or 64-bit DLL (binary) at https://github.com/hrgraf/pizmidi/releases.
Copy the DLL and the .ini file to your VST Plug-in directory (for 64-bit e.g. to C:\Program Files\VSTPlugins).
Use your favorite DAW, or any VST host, e.g. the free VSTHost or SAVIHost.


savihost/history.txt
SAVIHost Version History
========================

Before V1.42, SAVIHost's history could only be found on KvR (see 
  http://www.kvraudio.com/product/savihost-by-hermann-seib/news
for details); since I can't expect that everyone knows this, I'll keep track
of the changes here, too.
You can mentally add "Lots of bug fixes and internal changes" to each list of
changes; I won't explicitly state that each time :-)

V1.42
=====

Release Date: 2013-06-25
Changes:
- Fixed incompatibility with ReaJS and other PlugIns that don't pass back a
  chunk before one's been loaded
- VST3 MIDI Controller -> Parameter ID remapping didn't work correctly
- added soft clipping
- added VST Module Architecture distributions
  (presumably purely an intellectual exercise, no necessity)
- added PlugIn Editor window support for VST Module Architecture PlugIns.
  NB: this is limited to the few VSTMA PlugIns I could find on the 'net.
  If you got others which use additional features of Steinberg's over-the-top
  XML UI specification, please let me know.

V1.41
=====

Release Date: 2013-01-07
Changes:
- MIDI Transformations from VSTHost have been added.
- VST3-capable programs use VST SDK 3.5 instead of 3.1.

V1.40
=====

Release Date: 2011-09-20
Changes:
- Incoming MIDI SysEx messages killed the x64 versions.
- A bad "optimization" killed MME Wave Input device usage.

V1.39
=====

Release Date: 2011-08-04
Changes:
- SSE2 bug cured that caused ghastly noise when the output to a 24-bit sound
  card needed clipping.
- "Linear Knob mode" implemented.
- SAVIHost now contains a VST3.1 host in 32-bit and 64-bit versions.

V1.38
=====

Release Date: 2011-07-14
Changes:
- SAVIHost now comes in 32-bit and 64-bit versions.

V1.37
=====

Release Date: 2011-04-04
Changes:
- Use of SSE(2) when possible and appropriate; doesn't make much difference
  in SAVIHost.

V1.36
=====

Release Date: 2010-11-29
Changes:
- Wave Player level and Recording level can be adjusted.
- 1st configured MIDI Input was not opened correctly upon program start -
  corrected.
- .fxb bank files are saved in version 2 format now; if this is not wanted,
  it can be changed by adding a DWORD setting in the registry:
   HKEY_CURRENT_USER\Software\Seib\<PlugIn>\Settings\BankSaveVersion=1.

V1.35
=====

Release Date: 2010-11-01
Changes:
- just bug fixes

V1.34
=====

Release Date: 2009-11-26
Changes:
- Wave player supports auto-stereo mode for mono files.
- Can be minimized to system tray.

V1.33
=====

Release Date: 2009-10-15
Changes:
- Can play .mp3 files (using external mpg123 package - see VSTHost
  documentation for details).

V1.32
=====

Release Date: 2008-01-17
Changes:
- DirectSound support added (works in all Windows versions except Vista).
- Added /nomme, /nodsound, /noasio command line flags to suppress the use
  of the corresponding wave I/O method for special environments.
- Added /noft ("No Fault Tolerance") flag for debugging environments; in this
  mode, SAVIHost doesn't catch any exception, which eases the detection of
  badly behaving PlugIns.

V1.31
=====

Release Date: 2007-07-19
Changes:
- Improved exception handling (a classic "Who cares?" under normal conditions).
- "Autosave Bank" built in; SAVIHost can save the current bank to
  PlugInName.savihost.fxb in the PlugIn's directory now,
  if it has write permission there.


V1.30
=====

Release Date: 2007-04-05
Changes:
- Bank/program loading made more flexible; if the format of a loaded bank
  doesn't match the PlugIn settings, a Yes/No message box allows to override
  the abort (special for new Waldorf Wave 2.V).
- Plugins that try to set the host's speed (like Tap!) can do that now.
- Various internal changes & corrections.

V1.29
=====

Release Date: 2006-10-30
Changes:
- Entering registration data in daHornet did not work; corrected.
- Menus in Zebra 2 did not work; corrected.


V1.28
=====

Release Date: 2006-10-02
Changes:
- Stereo Input/Output ASIO Channel Selection corrected.
- Unused ASIO drivers should be released now.
- BPM changes can be directly entered on the main toolbar now.
- "Player Sync" setting finally provides a simple transport control
  (i.e. the hosts can now tell the plugins that they are playing at
   a certain position); makes it easier to run sequencer/tracker plugins
   in an orderly fashion.
- Finer-grained Priority selection.
- Wave Player files can be unloaded.


V1.27
=====

Release Date: 2006-06-12
Changes:
- When no Wave Input device is loaded, 2 channels of silence are used.
- I/O Remapping shows more meaningful device channel names.
- Wave Player had a problem if used without a Wave Input device; corrected.
- Wave Player has separate Autorepeat settings for playing and rewinding now.
- PC keyboard can be captured and sent to simulated MIDI keyboard now, even if
  this doesn't have the focus (or is hidden); can be toggled with F11.
- MIDI Panic button added.
- Stereo Input/Output ASIO Channel selection works again.

V1.26
=====

Release Date: 2006-05-08
Changes:
- Wave Player can now be inserted before or after the PlugIns; that means that
  you can send wave files through effects.
- Wave Player Auto-repeat introduced a little delay; corrected.
- Load / save file selection from PlugIns sometimes failed; corrected.
- Sending MIDI messages sometimes caused problems; corrected.

V1.25
=====

Release Date: 2006-02-06
Changes:
- VST SDK 2.4 support
  NB: while the programs should support all plugins written to the new
  specification, they do NOT call the 64-bit processDoubleReplacing() method,
  even if it's available.
  The reason is simply that the internal signal processing is done with
  32-bit floats; it would mean senseless conversions to and from 64-bit to
  cover the 64-bit functionality. VSTHost and SAVIHost are designed to run
  on older and slower machines, too, and I don't want to place heavy burdens
  on these by doubling all buffer sizes for a negligible improvement in sound
  quality.

V1.24
=====

Release Date: 2005-12-12
Changes:

common/pizvstbase.h
#ifndef PIZVSTBASE_H
#define PIZVSTBASE_H

#include "public.sdk/source/vst2.x/audioeffectx.h"
#include <iostream>
#include <vector>
#include "CVSTHost.h"

#ifdef _WIN32
#include <windows.h>
#include <Shlobj.h>
#endif

#ifdef _DEBUG
#define dbg(i)  { std::cout << i << "\n"; }
#define dbg2(i) { std::cout << i; }
#else 
#define dbg(i)  {}
#define dbg2(i) {}
#endif


typedef std::vector<VstMidiEvent> VstMidiEventVec;
typedef std::vector<VstMidiSysexEvent> VstSysexEventVec;

#define MAX_EVENTS_PER_TIMESLICE 256

struct MyVstEvents
{
    VstInt32 numEvents;
    VstIntPtr reserved;
    VstEvent* events[MAX_EVENTS_PER_TIMESLICE];
};


bool getInstancePath( char* outInstancePath, char* fileName, bool hostpath=true );
bool getAppDataPath( char* path, const char* name );

inline void removeExtension(char* filename) {
    if (!filename)
		return;
    char *lastdot = strrchr (filename, '.');
    if (lastdot)
        *lastdot = '\0';
}

inline void getHostStuff(char* host, bool &inst, int &numoutputs, bool &ignoreDefault) {
	ignoreDefault=false;
	dbg("host vendor: " << host);
	if (strncmp("Ableton",host,7)==0
		|| strncmp("Steinberg",host,9)==0
		|| strncmp("Twelve Tone",host,11)==0
		|| strncmp("Image-Line",host,10)==0
		|| strncmp("brainspawn",host,10)==0
		) {
		inst=true;
		numoutputs=2;
		ignoreDefault=true;
	}
	else if (strncmp("Polac",host,5)==0) {
		inst=true;
		numoutputs=0;
		ignoreDefault=true;
	}
	else if (strncmp("Native Instruments",host,18)==0
		|| strncmp("Music OS",host,8)==0) {
		inst=false;
		numoutputs=2;
		ignoreDefault=true;
	}
	else if (strncmp("PreSonus",host,8)==0) {
		inst=true;
		numoutputs=2;
		ignoreDefault=true;
	}
	else if (strncmp("Pedalboard 2",host,12)==0
		||   strncmp("Juce VST Host",host,13)==0
		||   strncmp("Plogue",host,6)==0
		||   strncmp("Ross Bencina",host,12)==0
		||   strncmp("MuTools.com",host,11)==0 
		||   strncmp("XT Software",host,11)==0 ) 
	{
		inst=false;
		numoutputs=0;
		ignoreDefault=true;
	}
}

inline void readIniFile(char* host, bool &inst, int &numinputs, int &numoutputs, int &bottomOctave, bool ignoreDefault) {
	dbg ("host product: " << host);
	char* path;
	char* name;
	path = new char[512];
	name = new char[512];
	getInstancePath(path,name,false);
	if (!strncmp(name,"midiAudioToCC",13)) inst=false;

	strcat(path,"pizmidi.ini");
	FILE *file = fopen(path,"r");
	if (file == NULL) {
		getAppDataPath(path,"pizmidi");
		strcat(path,"pizmidi.ini");
		file = fopen(path,"r");
	}
	if (file != NULL)
	{
		char line[128] = " ";
		bool hostmatch=false;
		char hostname[128] = "host=";
		strcat(hostname,host);
		strcat(hostname,"\n");
		dbg("====== pizmidi.ini =======");
		while ( fgets ( line, sizeof line, file ) != NULL )
		{
			dbg2(line);
			if (line[0]!=';') {
				if (!strncmp("default",line,7) && !ignoreDefault) {
					hostmatch=true;
				}
				else if (!strcmp(hostname,line)) {
					hostmatch=true;
				}
				else if (!strncmp("host=",line,5)) {
					hostmatch=false;
				}
				if (hostmatch) {
					if (strncmp("force effect",line,12)==0) inst=false;
					if (strncmp("force inst",line,10)==0) inst=true;
					if (strncmp("audio outs",line,10)==0) numoutputs=2;
					if (strncmp("no audio outs",line,13)==0) numoutputs=0;
					if (strncmp("audio ins",line,9)==0) numinputs=2;
					if (strncmp("no audio ins",line,12)==0) numinputs=0;
					if (strncmp("bottom octave=",line,14)==0) {
						bottomOctave=atoi((const char*)(&line[14]));
					}
				}
			}
		}
		dbg("====== end of pizmidi.ini =======\n");		
		fclose ( file );
	}
	char* pch = strchr(name,'_');
	if (pch) {
		if (!strncmp("_inst.",pch,6)) {
			inst=true;
			numoutputs=2;
		}
		else if (!strncmp("_nofx.",pch,6)) {
			inst=true;
			numoutputs=2;
		}
		else if (!strncmp("_fx.",pch,4)) {
			inst=false;
			numinputs=2;
			numoutputs=2;
		}
	}
	dbg ("pizmidi.ini path: " << path);
	dbg ((inst?"instrument":"effect"));
	dbg ("ins: " << numinputs);
	dbg ("outs: " << numoutputs);
	delete path;
	delete name;
}


inline bool readDefaultBank(char* plugName, CFxBank *bank) {
	char* path;
	char* adpath;
	char* name;
	path = new char[512];
	adpath = new char[512];
	name = new char[512];
	getInstancePath(path,name,false);
	strcat(name,".fxb");
	strcat(path,name);
	if (getAppDataPath(adpath,"pizmidi"))
	{
		strcat(adpath,name);
		FILE *file = fopen ( adpath, "r" );
		if ( file != NULL ) {
			strcpy(path,adpath);
		}
	}
	bank->LoadBank(path);
	if (bank->LoadBank(path)) return true;
	return false;
}

inline bool readDefaultBank(char* plugName, CFxBank *bank, VstInt32 ID) {
	char* path;
	char* name;
	path = new char[512];
	name = new char[512];
	if (!getAppDataPath(path,"pizmidi"))
		getInstancePath(path,name,false);
	strcat(path,plugName);
	strcat(path,".fxb");
	bank->LoadBank(path);
	if (bank->LoadBank(path)) {
		return (bank->GetFxID()==ID);
	}
	return false;
}

#endif
